src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs

```diff
@@ -0,0 +1,191 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.UnusedReferences
+{
+    internal static class UnusedReferencesService
+    {
+        // This is the order that we look for used references. We set this processing order because we
+        // want to favor transitive references when possible. For instance we process Projects before
+        // Packages, since a particular Package could be brought in transitively by a Project reference.
```

Comment:
> In this case we can either drop the project reference or drop the reference of B to P, and we're choosing to keep the project reference?

You make a good point. Seems like it would make more sesne to have two passes, first pass looking for direct references and second pass looking for transitive.